scripts/mod_loader/modapi/vector.lua

```diff
@@ -0,0 +1,380 @@
+
+math.vec2 = {}
+math.vec3 = {}
+math.vec4 = {}
+
+local mt_vec2 = {
+	x = 1, y = 2,
+	width = 1, height = 2,
+	w = 1, h = 2,
+	top = 1, bottom = 1, left = 2, right = 2,
+	t = 1, b = 1, l = 2, r = 2,
+	[1] = 1, [2] = 2
+}
+
+local mt_vec3 = {
+	x = 1, y = 2, z = 3,
+	i = 1, j = 2, k = 3,
+	[1] = 1, [2] = 2, [3] = 3
+}
+
+local mt_vec4 = {
+	north = 1, south = 2, west = 3, east = 4,
+	top = 1, bottom = 2, left = 3, right = 4,
+	t = 1, b = 2, l = 3, r = 4,
+	
+	x = 1, y = 2, w = 3, h = 4, width = 3, height = 4,
+	[1] = 1, [2] = 2, [3] = 3, [4] = 4
+}
+
+local function traceback()
+	return Assert.Traceback and debug.traceback("\n", 3) or ""
+end
+
+local function AssertIsVector(name, expected, actual, msg)
+	msg = (msg and msg .. ": ") or ""
+	msg = msg .. string.format("Expected %s, but was %s", name, type(actual), traceback())
+	assert(getmetatable(actual) == expected, msg)
+end
+
+local function AssertIsVector2(actual, msg)
+	AssertIsVector("math.vec2", mt_vec2, actual, msg)
+end
+
+local function AssertIsVector3(actual, msg)
+	AssertIsVector("math.vec3", mt_vec3, actual, msg)
+end
+
+local function AssertIsVector4(actual, msg)
+	AssertIsVector("math.vec4", mt_vec4, actual, msg)
+end
+
+local function addVectors(a, b)
+	
+	if #a == 4 or #b == 4 then
+		return math.vec4(
+			a[1] + b[1],
+			a[2] + b[2],
+			a[3] + b[3],
+			a[4] + b[4]
+		)
+		
+	elseif #a == 3 or #b == 3 then
+		return math.vec3(
+			a[1] + b[1],
+			a[2] + b[2],
+			a[3] + b[3]
+		)
+		
+	elseif #a == 2 or #b == 2 then
+		return math.vec2(
+			a[1] + b[1],
+			a[2] + b[2]
+		)
+	end
+	
+	return nil
```

Comment:
Maybe throw an error? It might be difficult for people to debug that they're trying to add two vectors of different sizes.

Reply:
If I am reading the code right, adding vectors of different sizes will return the larger size. This nil case means you had vectors of sizes less than 2 or greater than 4.

Reply:
Ah, you're right - I blindly assumed those would be `and` guards, not `or` guards

Reply:
I guess throwing error wouldn't hurt either way.

Reply:
Changed `return nil` to throw an error instead